### PR TITLE
[GEOS-9719] GeoJSONPPIO should use global number of decimals setting

### PR DESCRIPTION
**Pull request by jnh5y**
Addresses the rounding issue for the GeoJSON WPS PPIO identified here:
https://gis.stackexchange.com/questions/272009/geoserver-increase-the-precision-of-the-json-return-of-wps-call

Signed-off-by: Jim Hughes <jnh5y@ccri.com>

<Include a few sentences describing the overall goals for this Pull Request. Please do not remove the checklist below, pull requests missing it will be ignored.>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Comment by jnh5y** [comment]
@ianturton Thanks for responding with enough of a suggestion on the GIS Exchange to help me and my colleague find and fix this issue quickly!  (We weren't the reporters; just happy to knock out a small thing like this!)

**Comment by jnh5y** [comment]
Ok, the unit tests I added rely on the global settings being constant during the test.  That means that the test cases can fail if they run concurrently.  Is there an example of sorting this out in other unit tests?

**Comment by aaime** [comment]
We don't really run same module tests concurrently. Even with a parallel build, each module gets its own private JVM and runs tests sequentially inside of it.

**Comment by jnh5y** [comment]
@aaime thanks.  Running the new test in IntelliJ worked fine.  Building from the command line with -T4 has been throwing errors.  

I'll try building without the -T locally.

**Comment by aaime** [comment]
In general, or for this particular PR? I normally run with -T6 without troubles (on a machine with 8 physical cores, a Ryzen 1700x)

**Comment by jnh5y** [comment]
It is happening on this PR with my new test.  I wonder if I'm missing something obvious in setting/resetting the global settings?

**Comment by jodygarnett** [comment]
> It is happening on this PR with my new test. I wonder if I'm missing something obvious in setting/resetting the global settings?

Testing the output functions is what is important, rather than your ability to manage settings. Can you refactor so this functionality can be tested by a simple unit test?

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ppio/GeoJSONPPIOTest.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        int dec = global.getSettings().getNumDecimals();
+        global.getSettings().setNumDecimals(4);
+        getGeoServer().save(global);
+
+        ByteArrayOutputStream os2 = new ByteArrayOutputStream(1024);
+        new GeoJSONPPIO.Geometries(gs).encode(point, os2);
+        String output2 = os2.toString();
+        assertEquals(output2, "{\"type\":\"Point\",\"coordinates\":[1.1235,2]}");
+
+        global.getSettings().setNumDecimals(dec);
```
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ppio/GeoJSONPPIOTest.java:81]: I would usually handle this in a try/finally block for extra safety...

Comment (jnh5y) [reply src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ppio/GeoJSONPPIOTest.java:81]: Even in a unit test?

**Comment by jnh5y** [comment]
> > It is happening on this PR with my new test. I wonder if I'm missing something obvious in setting/resetting the global settings?
> 
> Testing the output functions is what is important, rather than your ability to manage settings. Can you refactor so this functionality can be tested by a simple unit test?

I didn't see an obvious way to do that.  In some sense, my change is about the 'system' working together.  My second commit addresses the situation in the unit test framework, so I think this is good to go!

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ppio/GeoJSONPPIOTest.java (hunk 1 above)
Comment (jodygarnett) [reply src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ppio/GeoJSONPPIOTest.java:81]: yeah, this was based on testing changes coordinate reference system configuration (so any failure produced so many other failures it was hard to track down.)

Keep in mind if you run into multi-threaded testing problems in the future.

**Comment by jodygarnett** [comment]
@jnh5y can we squash this into a single commit, your second commit does not have the [GEOS-9719] in the commit message.

**Comment by jnh5y** [comment]
@jodygarnett thanks for the review and approval!

**Pull request merged**